daemon/device_unix_test.go

```diff
@@ -0,0 +1,152 @@
+}
+
+// Based on test from runc:
+// https://github.com/opencontainers/runc/blob/v1.0.0/libcontainer/devices/device_unix_test.go#L34-L47
+func TestHostDevicesOsReaddirFailure(t *testing.T) {
+	testError := fmt.Errorf("test error: %w", os.ErrPermission)
+
+	// Override os.ReadDir to inject error.
+	osReaddir = func(dirname string) ([]fs.DirEntry, error) {
+		return nil, testError
+	}
+	// Override userns.RunningInUserNS to ensure not running in user namespace.
+	usernsRunningInUserNS = func() bool {
+		return false
+	}
+	defer cleanupTest()
+
+	_, err := HostDevices()
+	if !errors.Is(err, testError) {
+		t.Fatalf("Unexpected error %v, expected %v", err, testError)
+	}
+}
+
+// Based on test from runc:
+// https://github.com/opencontainers/runc/blob/v1.0.0/libcontainer/devices/device_unix_test.go#L34-L47
+func TestHostDevicesOsReaddirFailureIfRunningInUserNS(t *testing.T) {
+	testError := fmt.Errorf("test error: %w", os.ErrPermission)
+
+	// Override os.ReadDir to inject error.
+	osReaddir = func(dirname string) ([]fs.DirEntry, error) {
+		return nil, testError
+	}
+	// Override userns.RunningInUserNS to ensure running in user namespace.
+	usernsRunningInUserNS = func() bool {
+		return true
+	}
+	defer cleanupTest()
+
+	_, err := HostDevices()
+	if !errors.Is(err, nil) {
+		t.Fatalf("Unexpected error %v, expected %v", err, nil)
+	}
+}
+
+// Based on test from runc:
+// https://github.com/opencontainers/runc/blob/v1.0.0/libcontainer/devices/device_unix_test.go#L49-L74
+func TestHostDevicesOsReaddirDeepFailure(t *testing.T) {
+	testError := fmt.Errorf("test error: %w", os.ErrPermission)
+	called := false
+
+	// Override os.ReadDir to inject error after the first call.
+	osReaddir = func(dirname string) ([]fs.DirEntry, error) {
+		if called {
+			return nil, testError
+		}
+		called = true
+
+		// Provoke a second call.
+		fi, err := os.Lstat("/tmp")
+		if err != nil {
+			t.Fatalf("Unexpected error %v", err)
+		}
+
+		return []fs.DirEntry{fs.FileInfoToDirEntry(fi)}, nil
+	}
+	// Override userns.RunningInUserNS to ensure not running in user namespace.
+	usernsRunningInUserNS = func() bool {
+		return false
+	}
+	defer cleanupTest()
+
+	_, err := HostDevices()
+	if !errors.Is(err, testError) {
+		t.Fatalf("Unexpected error %v, expected %v", err, testError)
+	}
+}
+
+// Based on test from runc:
+// https://github.com/opencontainers/runc/blob/v1.0.0/libcontainer/devices/device_unix_test.go#L49-L74
+func TestHostDevicesOsReaddirDeepFailureIfRunningInUserNS(t *testing.T) {
+	testError := fmt.Errorf("test error: %w", os.ErrPermission)
+	called := false
+
+	// Override os.ReadDir to inject error after the first call.
+	osReaddir = func(dirname string) ([]fs.DirEntry, error) {
+		if called {
+			return nil, testError
+		}
+		called = true
+
+		// Provoke a second call.
+		fi, err := os.Lstat("/tmp")
+		if err != nil {
+			t.Fatalf("Unexpected error %v", err)
+		}
+
+		return []fs.DirEntry{fs.FileInfoToDirEntry(fi)}, nil
```

Comment:
Just a note; this won't work on Go 1.16, but we currently don't run that version in CI; given that this only affects testing code (and master moved to go1.17), perhaps that's not a problem; we added code in https://github.com/moby/moby/pull/42797 to work around that, but not sure if it's worth going through that to make the test work on Go 1.16
(Perhaps a build-tag for this file _could_ work as a quick fix)

Reply:
I added a build tag. In `runc` there is a similar workaround, too. Most probably we are going to move the missing unit tests to `containerd` anyway.